File: provider-swift/Sources/ProviderCore/Protocol/ModelAutopilot.swift
import Foundation

/// Protocol 1 is cached-only. Absent snapshot means no consent; neither an empty
/// enabled_models allowlist nor an advertised build implies opt-in.
public struct ModelAutopilotSnapshot: Codable, Sendable, Equatable {
    public var protocolVersion: Int = 1
    public var enabled: Bool
    public var cachedOnly: Bool = true
    public var minDwellSeconds: Int
    public var pinnedModels: [String]
    public var maxModelSlots: Int
    public var residentModels: [ModelAutopilotResident]
    /// Weight-only headroom while retaining ALL residents, from the load gate.
    public var freeForLoadNoEvictGb: Double?
    public var activeCommandId: String?
    public var lastCommandId: String?
    public var lastCommandStatus: ModelAutopilotStatus.State?

    public init(enabled: Bool, minDwellSeconds: Int = 1_800, pinnedModels: [String] = [],
                maxModelSlots: Int = 1, residentModels: [ModelAutopilotResident] = [],
                freeForLoadNoEvictGb: Double? = nil, activeCommandId: String? = nil,
                lastCommandId: String? = nil, lastCommandStatus: ModelAutopilotStatus.State? = nil) {
        self.enabled = enabled
        self.minDwellSeconds = minDwellSeconds
        self.pinnedModels = pinnedModels
        self.maxModelSlots = maxModelSlots
        self.residentModels = residentModels
        self.freeForLoadNoEvictGb = freeForLoadNoEvictGb
        self.activeCommandId = activeCommandId
        self.lastCommandId = lastCommandId
        self.lastCommandStatus = lastCommandStatus
    }

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol", enabled
        case cachedOnly = "cached_only", minDwellSeconds = "min_dwell_seconds"
        case pinnedModels = "pinned_models", maxModelSlots = "max_model_slots"
        case residentModels = "resident_models", freeForLoadNoEvictGb = "free_for_load_no_evict_gb"
        case activeCommandId = "active_command_id", lastCommandId = "last_command_id"
        case lastCommandStatus = "last_command_status"
    }
}

public struct ModelAutopilotResident: Codable, Sendable, Equatable {
    public var modelId: String
    public var residentSeconds: Int
    public var idleSeconds: Int
    /// Scanner-padded load estimate; NOT extra free memory or throughput.
    public var weightsGb: Double
    /// Actual slot weight ownership in GiB, the only victim reclaim credit.
    public var residentGb: Double?
    public init(modelId: String, residentSeconds: Int, idleSeconds: Int, weightsGb: Double,
                residentGb: Double? = nil) {
        self.modelId = modelId
        self.residentSeconds = residentSeconds
        self.idleSeconds = idleSeconds
        self.weightsGb = weightsGb
        self.residentGb = residentGb
    }
    enum CodingKeys: String, CodingKey {
        case modelId = "model_id", residentSeconds = "resident_seconds"
        case idleSeconds = "idle_seconds", weightsGb = "weights_gb"
        case residentGb = "resident_gb"
    }
}

public struct ModelAutopilotCommand: Codable, Sendable, Equatable {
    public var commandId: String
    public var loadModelId: String?
    public var unloadModelIds: [String]
    public var expectedResidentModels: [String]
    public var expiresAtMs: Int64
    public var leaseSeconds: Int

    public init(commandId: String, loadModelId: String? = nil, unloadModelIds: [String] = [],
                expectedResidentModels: [String] = [], expiresAtMs: Int64, leaseSeconds: Int = 1_800) {
        self.commandId = commandId
        self.loadModelId = loadModelId
        self.unloadModelIds = unloadModelIds
        self.expectedResidentModels = expectedResidentModels
        self.expiresAtMs = expiresAtMs
        self.leaseSeconds = leaseSeconds
    }
    enum CodingKeys: String, CodingKey {
        case commandId = "command_id", loadModelId = "load_model_id"
        case unloadModelIds = "unload_model_ids", expectedResidentModels = "expected_resident_models"
        case expiresAtMs = "expires_at_ms", leaseSeconds = "lease_seconds"
    }
}

public struct ModelAutopilotStatus: Codable, Sendable, Equatable {
    public enum State: String, Codable, Sendable { case started, succeeded, failed }
    public var commandId: String
    public var status: State
    public var error: String?
    public var modelAutopilot: ModelAutopilotSnapshot
    public init(commandId: String, status: State, error: String? = nil, modelAutopilot: ModelAutopilotSnapshot) {
        self.commandId = commandId
        self.status = status
        self.error = error
        self.modelAutopilot = modelAutopilot
    }
    enum CodingKeys: String, CodingKey {
        case commandId = "command_id", status, error
        case modelAutopilot = "model_autopilot"
    }
}

File: coordinator/protocol/model_autopilot.go
package protocol

// Autopilot is an explicit, connection-scoped opt-in. Missing snapshots mean
// disabled; protocol negotiation never relies on a provider version string.
const (
	TypeModelAutopilot       = "model_autopilot"
	TypeModelAutopilotStatus = "model_autopilot_status"
	ModelAutopilotProtocol   = 1
)

type ModelAutopilotResident struct {
	ModelID         string   `json:"model_id"`
	ResidentSeconds float64  `json:"resident_seconds"`
	IdleSeconds     float64  `json:"idle_seconds"`
	WeightsGB       float64  `json:"weights_gb"`
	ResidentGB      *float64 `json:"resident_gb,omitempty"`
}

type ModelAutopilotState struct {
	Protocol             int                      `json:"protocol"`
	Enabled              bool                     `json:"enabled"`
	CachedOnly           bool                     `json:"cached_only"`
	MinDwellSeconds      int                      `json:"min_dwell_seconds"`
	PinnedModels         []string                 `json:"pinned_models"`
	MaxModelSlots        int                      `json:"max_model_slots"`
	ResidentModels       []ModelAutopilotResident `json:"resident_models"`
	FreeForLoadNoEvictGB *float64                 `json:"free_for_load_no_evict_gb,omitempty"`
	ActiveCommandID      string                   `json:"active_command_id,omitempty"`
	LastCommandID        string                   `json:"last_command_id,omitempty"`
	LastCommandStatus    string                   `json:"last_command_status,omitempty"`
}

// A command grants permission to unload ONLY the named victims. The provider
// must validate the full resident-set precondition, local work, dwell, pins,
// slot/memory feasibility and expiration before mutating residency. It must
// never invoke implicit LRU eviction to make this operation fit.
type ModelAutopilotMessage struct {
	Type                   string   `json:"type"`
	CommandID              string   `json:"command_id"`
	LoadModelID            string   `json:"load_model_id,omitempty"`
	UnloadModelIDs         []string `json:"unload_model_ids"`
	ExpectedResidentModels []string `json:"expected_resident_models"`
	ExpiresAtMS            int64    `json:"expires_at_ms"`
	LeaseSeconds           int      `json:"lease_seconds"`
}

// Status is an acknowledgement, not authoritative scheduler capacity. Only a
// fresh sequenced heartbeat can reconcile the resulting BackendCapacity slots.
type ModelAutopilotStatusMessage struct {
	Type           string               `json:"type"`
	CommandID      string               `json:"command_id"`
	Status         string               `json:"status"`
	Error          string               `json:"error,omitempty"`
	ModelAutopilot *ModelAutopilotState `json:"model_autopilot,omitempty"`
}
